Stores token source metadata in the service description

shamsimam opened the pull request:
## Changes proposed in this PR

We are adding support for the tokens (and their versions) used to create a service description in this PR. The field is stored as the `source-tokens` metadata field in the service description. While computing the service description, we auto-populate this new metadata field. The metadata field also needs to be populated during the waiter consent workflow.

<img width="562" alt="source-tokens" src="https://user-images.githubusercontent.com/6611249/40184979-2a9b1386-59b7-11e8-9a8a-a8d58f0828d0.png">

As a result, we now have a distinction between service parameters and service description (includes service parameters and service metadata). This distinction also carries forward to tokens which only specify service parameter values.

### Summary of changes
- introduces `service-parameter-keys`, `service-non-override-keys` and `service-metadata-keys`
- distinguishes between token `service-description-template` and `service-parameter-template`

**Note**: A result of these changes is the refactoring to rename `service-description-keys` to `service-parameter-keys` and (in the context of tokens) `service-description-template` to `service-parameter-template`. Let me know if you prefer these renames to be provided in two separate PRs to minify the diff of this PR.

## Why are we making these changes?

Our long-term goal is to be able to GC outdated services quicker. This PR enables that feature by providing the token information used to create a service. As a result of this change, on-the-fly services with the same service parameters as a tokenized service will be treated as two different services.

sradack commented:
Please rebase.  

A couple of things before I dive into the details:

1. Having the "E" in the version hash (what we also use for ETag) now seems even more awkward to me.  "E" standing for "Etag" is an HTTP concept, and has bled into our service descriptions.
2. Should `source-tokens` be a vector instead of a map?  Tokens are provided in an ordered list, and using a map means we lose that order.

shamsimam commented:
Rebased.

> Having the "E" in the version hash (what we also use for ETag) now seems even more awkward to me. "E" standing for "Etag" is an HTTP concept, and has bled into our service descriptions.

The `E-` is a prefix we use to quickly identify token hashes, since it is an internal implementation detail we can change it whenever we like.

> Should source-tokens be a vector instead of a map? Tokens are provided in an ordered list, and using a map means we lose that order.

I changed it to a vector.

sradack commented on waiter/src/waiter/service_description.clj:
```diff
@@ -98,23 +98,30 @@
+  (set/union service-override-keys service-non-override-keys))
+
+; keys allowed in service description metadata, these need to be distinct from service parameter keys
+(def ^:const service-metadata-keys #{"source-tokens"})
```
How are these different from the "other" metadata keys?

shamsimam replied:
Currently all the fields we store in the service description are the service parameters, `surce-tokens` will be the first of its kind - a metadata entry that will be system generated, that we will not allow end users to configure.

sradack commented on waiter/src/waiter/service_description.clj:
```diff
@@ -498,10 +506,24 @@
       token-data->token-description))
 
+(defn token->service-parameter-template
+  "Retrieves the service description template for the given token containing only the service parameters."
```
What's the difference between a service description template and a service parameter template?

shamsimam replied:
Service description == service parameters + system generated metadata

sradack commented on waiter/src/waiter/service_description.clj:
```diff
@@ -791,14 +817,9 @@
-                                   :nil-on-missing? nil-on-missing?)]
-    (cond-> core-service-description
-            effective? (default-and-override metric-group-mappings kv-store service-description-defaults service-id))))
+   {:keys [effective? nil-on-missing? refresh] :or {effective? true nil-on-missing? true refresh false}}]
```
Please move this to a separate PR.

shamsimam replied:
Done.

sradack commented on waiter/src/waiter/token.clj:
```diff
@@ -429,11 +429,12 @@
-                                  :headers {"etag" (-> {:service-description-template new-service-description-template
-                                                        :token-metadata new-token-metadata}
-                                                       token-description->token-hash)})))))
+        (utils/map->json-response
```
Please move this to a separate PR.

sradack commented on waiter/src/waiter/service_description.clj:
```diff
@@ -746,7 +772,7 @@
         (throw (ex-info (utils/message :cannot-identify-service)
                         (error-message-map-fn passthrough-headers waiter-headers))))
       (sling/try+
-        (let [{:keys [service-id service-description core-service-description]}
```
Please move this to a separate PR.

shamsimam replied:
Done.

shamsimam commented:
Rebased.
